internal/uvm/vnuma.go

```
@@ -0,0 +1,166 @@
+// - `PhysicalNumaNodes` values at index `i` will be mapped to virtual node number `i`
+// - client is responsible for setting wildcard physical node numbers
+func prepareVNumaTopology(opts *Options) (*hcsschema.Numa, *hcsschema.NumaProcessors, error) {
+	if opts.MaxProcessorsPerNumaNode == 0 && len(opts.NumaMappedPhysicalNodes) == 0 {
+		// vNUMA settings are missing, return empty topology
+		return nil, nil, nil
+	}
+
+	var preferredNumaNodes []int64
+	for _, pn := range opts.PreferredPhysicalNumaNodes {
+		preferredNumaNodes = append(preferredNumaNodes, int64(pn))
+	}
+
+	build := osversion.Get().Build
+
+	// Implicit vNUMA topology.
+	if opts.MaxProcessorsPerNumaNode > 0 {
+		if build < osversion.V25H1Server {
+			return nil, nil, fmt.Errorf("implicit vNUMA topology is not supported on %d version of Windows", build)
+		}
+		if opts.MaxSizePerNode == 0 {
+			return nil, nil, fmt.Errorf("max size per node must be set when max processors per numa node is set")
+		}
+		numaProcessors := &hcsschema.NumaProcessors{
+			CountPerNode: hcsschema.Range{
+				Max: opts.MaxProcessorsPerNumaNode,
+			},
+		}
+		numa := &hcsschema.Numa{
+			MaxSizePerNode:         opts.MaxSizePerNode,
+			PreferredPhysicalNodes: preferredNumaNodes,
+		}
+		return numa, numaProcessors, nil
+	}
+
+	// Explicit vNUMA topology.
+
+	numaNodeCount := len(opts.NumaMappedPhysicalNodes)
+	if numaNodeCount != len(opts.NumaProcessorCounts) || numaNodeCount != len(opts.NumaMemoryBlocksCounts) {
+		return nil, nil, fmt.Errorf("mismatch in number of physical numa nodes and the corresponding processor and memory blocks count")
+	}
+
+	if build < osversion.V25H1Server {
+		return nil, nil, fmt.Errorf("explicit vNUMA topology is not supported on %d version of Windows", build)
+	}
+
+	numa := &hcsschema.Numa{
+		VirtualNodeCount:       uint8(numaNodeCount),
+		Settings:               []hcsschema.NumaSetting{},
+		PreferredPhysicalNodes: preferredNumaNodes,
```

duplicating my comment from a different comment thread, since the code I originally commented on has been removed:

I've confirmed that 1 and 2 work as one would expect: the platform will honor preferred physical nodes when creating an implicit vNUMA topology or when assigning pNUMA indices to wildcard/invalid physical node numbers.

3 is actually not a supported scenario and the validation code handles it.

That being said though, I think we should not support explicit placement + preferred physical nodes in the initial implementation and see if anyone would actually need this semi-implicit placement in the future. This means that we should return an error when:

- preferred physical nodes is used together with mapped physical nodes
- mapped physical nodes contain a wildcard/invalid node number (0xFF)

---

Lets discuss this offline and address in a follow up PR if needed